mkosi/backend.py

```diff
@@ -790,13 +790,16 @@ def spawn(
 def run(
     cmdline: Sequence[PathString],
     check: bool = True,
+    shell: bool = False,
     delay_interrupt: bool = True,
     stdout: _FILE = None,
     stderr: _FILE = None,
     env: Mapping[str, Any] = {},
     **kwargs: Any,
 ) -> CompletedProcess:
-    cmdline = [str(x) for x in cmdline]
+
+    if not shell:
+        cmdline = [str(x) for x in cmdline]
```

Comment:
Yes, and with this we should not hit the case where we pass a list with `shell=True`

Reply:
Ah, sorry, I was misunderstanding this, since I'm mostly only used to passing lists to `subproces.run` and that's what usually (always?) happening in the rest of the codebase.

How about?
```suggestion
    cmdline = [str(x) for x in cmdline]
    if shell and isinstance(cmdline, list):
        cmdline = " ".join(t for t in cmdline))
```

Normally I'd throw in a `shlex.quote` for good measure, too, but this would preclude the use with globs that you want.

Reply:
`isinstance(cmdline, list)` would always be true. I don't see much use in joining the list again when we get the string itself passed as a parameter.

Reply:
Sorry, I didn't properly explain my thinking. 

I quickly grepped through the codebase and didn't find  an instance of `run` not being called with a list as an argument. I might have missed something, but the vast majority of call sites are like this. This can't be really encoded on the type level, though, since the `cmdline` argument for `run` is `Sequence[str]` (it's actually not, but I want to simplify this), which was chosen because sequences are immutable by default and it's nice to be general about what one accepts (also it's hard to just change this to list, because lists are invariante and sequenes are covariant so mypy complains about this). 

Unfortunately this is a bit of a footgun, because `str` is also a `Sequence[str]`, so this doesn't ensure the assumption of something that's a bag of words coming in.

Long story short: I was always thinking about this with a list going in, and in that case the join is needed to make `shell=True` work. 

You're right about the `isinstance` there being pointless, I tacked that on without thinking, because I wanted to make the case of a string being passed in work, but in any case keeping a list as input for `run` makes the most sense overall.